Geometry shader emulation for macOS

gdkchan opened the pull request:
Overview
---

Some platforms, such as macOS, do not have geometry shader support at all. For those cases, we need to emulate the missing stage, either using a compute or separate vertex shader draw with rasterizer discard. This change implements geometry shader emulation using compute shader, which requires converting both the vertex and geometry stages to compute, 

Implementation
---

This implementation, unlike the one in macos1, uses compute shaders rather than vertex shaders. There are some advantages in doing that, such as being able to use subgroup operations (which are not supported on vertex shaders on macOS). It also has disadvantages, like not being able to easily fetch vertex buffer data.

**Vertex stage:**

Because the vertex shader runs before the geometry, we need to somehow run it before the geometry one runs, so it must also be done separately from the main draw, and this implementation will also do it on compute.

Compute shaders can't access vertex buffers, so instead they are bound as buffer textures. This has the advantage that *most* vertex formats are also supported as buffer texture formats. There are a few formats that are not supported, and in those cases, they are emulated using a format that is similar but has different component type (for example, some signed formats are emulated wth unsigned ones, and the sign extension is done on the shader). The implementation with buffer textures also happens to share the same limitations that the Vulkan spec has for vertex attributes: All offsets must be at least aligned to the format component size, or in case of packed formats, to the full packed size of the format. That is because for non-packed formats, the current implementation uses a single component. For example, RGB16 will be bound as R16, and the shader will add an offset for the actual component it needs to access (R, G or B). Since the format is R16, the vertex buffer offset must be aligned to 2 bytes.

The index buffer is also bound as a buffer texture. Like with vertex buffers, this has the advantage that we can easily control the format, based on the index type. For Ubyte index buffers, we set the format to R8Uint, for Ushort it is set to R16Uint, and for Uint it is set to R32Uint, there is no need for shader specialization.

The process for accessing a vertex input on the shader goes more or less like this:
- Read the vertex ID for the current instance from the index buffer, also calculate the vertex ID for the current instance.
- Read from the vertex info whether the respective vertex buffer is instanced.
- Read value from the buffer texture:
 - If the vertex buffer is instanced, use current instance vertex ID, otherwise use the one from the index buffer.
 - Add first vertex or first instance offsets (depending on the vertex buffer being instanced or not).
 - Multiply stride, if the format is not packed, add the component offset.
 - If the format is not packed, check if the accessed component actually exists on the format. If it doesn't, the value is forced to 0 (for R/G/B) or 1 (for A).
 
The vertex shader built-ins that doesn't exist on compute are replaced with equivalents:
- `gl_VertexID`/`gl_VertexIndex` is the value read from the index buffer for the current invocation (`gl_GlobalInvocationID.x`), with the base vertex added.
- `gl_InstanceID`/`gl_InstanceIndex` becomes `gl_GlobalInvocationID.y` subtracted from base instance if needed.
- `gl_BaseVertex`/`gl_BaseInstance` is read from the vertex info buffer.

All output are first stored into a temporary "local memory", which is jsut an array. At the end of the shader, all written outputs are written into a output storage buffer.

**Geometry stage:**

All geometry shader inputs are replaced with loads from the storage buffer that is written by the compute vertex stage. Because the vertex stage does not do any topology conversion, the vertex data will be in the same order of the index buffer. However, the geometry stage is meant to access full primitives. So if for example, the topology is "triangle strip", It is necessary to ensure the first  geometry shader invocation will access vertices 0 1 2, the 2nd one 2 1 3, etc. For that, a buffer texture called "topology remap" buffer is used. It remaps the linear indices that the geometry shader access (0 1 2 3 4 5) into the correct indices for the current topology (0 1 2 3 4 5 for triangles, 0 1 2 2 1 3 for triangle strip, 0 1 2 0 2 3 for triangle fan, etc).

Just like with the vertex stage, all output are stored into a temporary "local memory". This also makes easy to implement output loads, as it can just load from this local memory without paying the cost of a storage buffer access for example.

There are two operations that are specific to geometry shader, and those are also replaced/expanded when converted to compute:
- `EmitVertex`: This stores all outputs into the storage buffer, and advances the "current vertex", in addition to writing the vertex ID into the index buffer, and also advancing the "current index".
- `EndPrimitive`: This stores the value -1 into the index buffer, in addition to advancing the "current index". -1 is the primitive restart value, is is useful on the draw operation that consumes the geometry shader outputs, primitive restart causes the current strip to end and starts a new one, which is what `EndPrimitive` does on a geometry shader.

Just like with the vertex stage, there are a few geometry shader specific built-ins that are replaced too.

The last step on a converted geometry shader is filling the remaining space on the output index buffer.  This is necessary because geometry shaders can produce an arbitrary amount of primitives. They only specify the *maximum* amount of vertices that can be produced, but the shader might produce less than the maximum, and in that case we need to ensure the remaining uninitialized state will not result in garbage being accessed later. All the buffers are allocated using the *maximum* output vertices count.

 **Fragment stage:**
 
 This is the last stage. It is done with an actual draw rather than compute dispatch. It has a vertex passthrough shader that will read the data from the last output buffer (from geometry, or vertex if there's no geometr shader), and pass it to fragment, which will finally render it into the output render targets.
 
If rasterizer discard is enabled, this step *might* not be necessary, and can be skipped. Note that this still needs to exist if the host supports transform feedback, for the case where the vertex passthrough stage does the transform feedback.

Current limitations
---

There are a few scenarios that are not currently supported. Some of them are not hard to implement, but I couldn't find any game that uses it, and I already spent a fair bit of time on this, plus it's generally better than was is on macos1,  so I consider it "good enough" for now.

The unsupported scenarios are:
- Converted draw with primitive restart: Requires linear pass over the index buffer, complicated to implement in a efficient way.
- Indirect draw: Not too complicated to implement, but will probably require using indirect dispatch as we don't know the draw counts.
- Vertex attributes/buffers with misaligned offsets: This is a violation of the Vulkan spec too, but we have some handling on the Vulkan backend. However there is no "re-striding" on draws emulated on compute, so they will be broken there if any game uses it.
- Primitives generated/transform feedback primitives written count: Metroid Prime Remastered needs this. But what uses it is not currently working properly in any platform anyway, so I figured it's not worth the time to implement right now.
- Tessellation + geometry shader is not supported at all. It would require emulating tessellation on compute, I'm not sure how hard that would be.
- Transform feedback from geometry shaders will not work properly, unless it always outputs the maximum amount of vertices. All the other existing limitations of transform feedback emulation also applies.
- Shader caching of converted shaders is not supported. Currently the shader cache only stores one "host" shader. Storing multiple will require a few changes. Most games should not have many (or any) converted shaders, so I think not caching them is fine for now.

Also, not really a limitation, but the current implementation will always do one or more compute dispatches, followed by a draw. That pattern is very inefficient and will cause noticeable performance degradation on games that uses geometry shaders on a lot of draws. We should try to batch the dispatch and draw operations to mitigate that, but this is not something simple to do.

Other notable changes
---

- The transform feedback info buffer has been removed. Now that information is stored on the support buffer.
- The code to move `gl_Layer` assignment from the geometry shader to the vertex shader for UE4 games has been removed. Now that we have geometry shader emulation, it is no longer needed (although moving it to vertex is more efficient, but since its only 1-2 draws per frame, it shouldn't really matter).

Affected games
---

This is not an exhaustive list, just some games I know that uses geometry shaders.

Note: The "before" screenshots were taken from the `master` branch. The cases where macos1 differs from this branch are noted on each comparison. If there is no note, is because the game renders/functions correctly on macos1 too.

**Nier Automata: The End of YoRHa edition:**

Before:
<img width="1392" alt="Captura de Tela 2023-08-10 às 21 50 13" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/13a64762-27db-41b3-aeb8-8fe56bc91d3a">
After:
<img width="1392" alt="Captura de Tela 2023-08-10 às 22 07 08" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/0ccceeff-6472-42c8-a9cc-b4cfff62736d">
<img width="1392" alt="Captura de Tela 2023-08-10 às 22 14 19" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/59759cad-32ee-4197-8bd8-6614f7f54f97">
I would still not consider it playable due to the low performance, and there are a few glitches remaining. Performance will likely improve considerably if we batch draws, or if MoltenVK implements geometry shaders using mesh shaders (but the latter would likely also not support subgroup operations, we would need to do something about this).

This game crashes before getting in-game on macos1.

**Marvel Ultimate Alliance 3:**

Before:
<img width="1392" alt="Captura de Tela 2023-08-10 às 23 00 29" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/6566860f-160a-4b96-9b6f-3a35fad75883">
After:
<img width="1392" alt="Captura de Tela 2023-08-10 às 22 29 23" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/b5b7094f-f766-4f28-9b55-8250a378166e">
Like Nier Automata, this also uses subgroup operations from geometry shaders, and has a lot of geometry shader draws. So it would likely have a large performance improvement from batching or using mesh shaders as mentioned above.

This does not render correctly on macos1.

**Splatoon 3:**

Before:
<img width="1392" alt="Captura de Tela 2023-08-10 às 21 58 51" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/46f0e4ce-5c4f-4fe4-ba47-412805160f49">
After:
<img width="1392" alt="Captura de Tela 2023-08-10 às 22 04 43" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/14345592-3132-491f-b3d6-61f0cbf47af8">
Pay attention to the "fuzzy ooze".

**Usotsuki Hime to Moumoku Ouji (The Liar Princess and the Blind Prince):**

Before:
<img width="1392" alt="Captura de Tela 2023-08-10 às 21 59 26" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/1bc478a9-8564-471d-bb5a-eb51ff901ae2">
After:
<img width="1392" alt="Captura de Tela 2023-08-10 às 22 01 19" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/fa25a520-707c-4ae0-8bfa-fbacabdac4e7">
This game uses geometry shaders for particle effects extensively. On the title screen, it can  be noticed on the music notes coming out of the wolf's mouth.

**Crash Bandicoot N. Sane Trilogy:**

Before:
<img width="1392" alt="Captura de Tela 2023-08-10 às 22 55 41" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/d57aea40-f245-42ce-8c9b-776a09492fd2">
After:
<img width="1392" alt="Captura de Tela 2023-08-10 às 22 24 41" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/f8275d4e-b53e-44bf-b56d-cca1d70de2b8">

**Shin Megami Tensei III:**

Before:
It crashes on master before getting in-game, so there's nothing to show. It's an error creating a pipeline before the vertex shader outputs and fragment inputs don't match (likely because it is supposed to have a geometry shader in-between that was removed).
After:
<img width="1392" alt="Captura de Tela 2023-08-10 às 22 37 02" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/be20d60d-0234-4c2d-91ad-997235de0934">

**Super Mario Maker 2:**

Before:
<img width="1392" alt="Captura de Tela 2023-08-10 às 22 48 29" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/0a6c2fe6-4844-4bd5-a56c-0c173aca586f">
After:
<img width="1392" alt="Captura de Tela 2023-08-10 às 22 46 13" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/4fb914a8-8411-4d38-a64b-60e3281b9a8b">

**Mario Strikers: Battle League:**

Before:
<img width="1392" alt="Captura de Tela 2023-08-10 às 22 51 50" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/918d3bd6-2b9e-41e6-8704-035d42324f8d">
<img width="1392" alt="Captura de Tela 2023-08-10 às 22 52 40" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/6668912a-2ddc-4063-a56c-ec5f59afb976">
After:
<img width="1392" alt="Captura de Tela 2023-08-10 às 22 39 58" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/2c1a5b26-37ee-40b6-87c4-91941d994aa3">
<img width="1392" alt="Captura de Tela 2023-08-10 às 22 40 51" src="https://github.com/Ryujinx/Ryujinx/assets/5624669/7b7f17ad-eb8a-4936-8432-09ecdda5773e">
There are 2 fixes here. The crowd now fully renders, and the strokers on the hyper strike animation also render. Note that the hyper strike animation was broken (in a different way) on macos1.

Future improvements
---

As mentioned before, performance improvements will come from batching draws or geometry shaders being implemented on MoltenVK. If geometry shaders are implemented on MoltenVK, the emulator will automatically use it and this compute conversion will be disabled, so there's nothing that needs to be done from our side to support it (other than updating to a newer MoltenVK version).

Contributes to #4062 (pretty much the only change missing at this point).
Fixes #4506.
Fixes #5583.
Fixes #5614.

siegmund-heiss-ich commented:
Just tried to test Metroid Prime Remastered, because there were problems with some colour artifacts without working geometry shaders.

The game got stuck right after the intro with these logged errors:
```
00:00:59.763 |E| .NET ThreadPool Worker [mvk-error] VK_ERROR_INVALID_SHADER_NV: Unable to convert SPIR-V to MSL:
00:00:59.763 |E| .NET ThreadPool Worker MSL conversion error: Vertex attribute type mismatch between host and shader
00:00:59.763 |E| .NET ThreadPool Worker 
00:00:59.763 |E| .NET ThreadPool Worker 
00:00:59.763 |E| .NET ThreadPool Worker [mvk-error] VK_ERROR_INVALID_SHADER_NV: Vertex shader function could not be compiled into pipeline. See previous logged error.
00:00:59.764 |E| .NET ThreadPool Worker Gpu : Background Compilation failed: Unexpected API error "ErrorInvalidShaderNV".
```

Hopefully this is helpful.

gdkchan commented:
> Just tried to test Metroid Prime Remastered, because there were problems with some colour artifacts without working geometry shaders.
> 
> The game got stuck right after the intro with these logged errors:
> 
> ```
> 00:00:59.763 |E| .NET ThreadPool Worker [mvk-error] VK_ERROR_INVALID_SHADER_NV: Unable to convert SPIR-V to MSL:
> 00:00:59.763 |E| .NET ThreadPool Worker MSL conversion error: Vertex attribute type mismatch between host and shader
> 00:00:59.763 |E| .NET ThreadPool Worker 
> 00:00:59.763 |E| .NET ThreadPool Worker 
> 00:00:59.763 |E| .NET ThreadPool Worker [mvk-error] VK_ERROR_INVALID_SHADER_NV: Vertex shader function could not be compiled into pipeline. See previous logged error.
> 00:00:59.764 |E| .NET ThreadPool Worker Gpu : Background Compilation failed: Unexpected API error "ErrorInvalidShaderNV".
> ```
> 
> Hopefully this is helpful.

I don't think this is a new error, as far I can tell, it is caused because the vertex shader uses input indexing (meaning it accesses inputs with an offset that is not constant, which indicates that it uses an array for the vertex shader inputs). Because it is an array, we can't set the correct type for each attribute and there is a type mismatch.

Anyway, this is not the cause of the freeze. The cause was the transform feedback on compute that I added. For some reason it was causing the game to hang on macOS. Since it was not really adding anything right now (the two games that requires it don't have transform feedback working properly for other reasons anyway) so I removed it, the game should be working again.

jneen commented:
Just built on this version to test on my m1 and can confirm it does seem to resolve #4506! I did get a crash, but I assume that's up to this being a WIP.

IuriPires commented:
Hey @gdkchan have you seen this? https://github.com/KhronosGroup/MoltenVK/pull/1815 maybe we gonna have native support (from MoltenVK)

gdkchan commented:
> Hey @gdkchan have you seen this? [KhronosGroup/MoltenVK#1815](https://github.com/KhronosGroup/MoltenVK/pull/1815) maybe we gonna have native support (from MoltenVK)

Yes, I have seen it, the PR has been open for a few months now. Also from their recent roadmap, it seems that their plans have changed, so geometry shader support on MoltenVK might be a long way off still. In any case, as noted on the PR, if MoltenVK receives geometry shader support and we update the library that we ship, the emulator should use the MoltenVK implementation without any change being required on the emulator side.

riperiperi commented on src/Ryujinx.Graphics.Shader/Translation/ShaderDefinitions.cs:
```diff
@@ -320,5 +319,35 @@ public AttributeType GetAttributeType(int location)
+
+        public bool IsAttributeSint(int location)
+        {
+            return (_graphicsState.AttributeTypes[location] & ~(AttributeType.Packed | AttributeType.PackedRgb10A2Signed)) == AttributeType.Sint;
```
Masking out these flags is a bit messy, maybe `AttributeType` should have an enum member that's `Packed | PackedRgb10A2Signed` so this isn't repeated.

riperiperi reviewed (approved):
Converting vtg shaders into compute is a totally crazy thing to do, and here it works perfectly. Fantastic work! I've tested a few games and they seem to work fine, though I don't actually have many games that use geometry shaders.

It's nice to see the geometry layer selection handled here instead of with the shader matching, that should work with all cases now.

I think it's a good idea to leave batching for a later PR, since it can turn into a whole can of worms, and I don't think anything here will be slower than `macos1` for the games that worked on that version.

riperiperi commented:
It's not exactly a normal application, but I did get this error when trying the citra retroarch core:

```
00:00:22.485 |W| GPU.MainThread Gpu Log: Shader translator: Invalid output "ClipDistance".
00:00:22.485 |W| GPU.MainThread Gpu Log: Shader translator: Invalid output "ClipDistance".
00:00:22.486 |W| GPU.MainThread Gpu Log: Shader translator: Invalid output "ClipDistance".
00:00:22.486 |W| GPU.MainThread Gpu Log: Shader translator: Invalid output "ClipDistance".
00:00:22.486 |W| GPU.MainThread Gpu Log: Shader translator: Invalid output "ClipDistance".
00:00:22.486 |W| GPU.MainThread Gpu Log: Shader translator: Invalid output "ClipDistance".
00:00:22.486 |W| GPU.MainThread Gpu Log: Shader translator: Invalid output "ClipDistance".
00:00:22.486 |W| GPU.MainThread Gpu Log: Shader translator: Invalid output "ClipDistance".
00:00:22.513 |E| GPU.MainThread Application : Unhandled exception caught: System.Collections.Generic.KeyNotFoundException: The given key 'Output.Position.0.0' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Ryujinx.Graphics.Shader.CodeGen.Spirv.Instructions.GetStoragePointer(CodeGenContext context, AstOperation operation, AggregateType& varType) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Shader/CodeGen/Spirv/Instructions.cs:line 1821
   at Ryujinx.Graphics.Shader.CodeGen.Spirv.Instructions.GenerateLoadOrStore(CodeGenContext context, AstOperation operation, Boolean isStore) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Shader/CodeGen/Spirv/Instructions.cs:line 1693
   at Ryujinx.Graphics.Shader.CodeGen.Spirv.Instructions.GenerateStore(CodeGenContext context, AstOperation operation) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Shader/CodeGen/Spirv/Instructions.cs:line 1153
   at Ryujinx.Graphics.Shader.CodeGen.Spirv.Instructions.Generate(CodeGenContext context, AstOperation operation) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Shader/CodeGen/Spirv/Instructions.cs:line 158
   at Ryujinx.Graphics.Shader.CodeGen.Spirv.SpirvGenerator.Generate(CodeGenContext context, AstBlock block) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Shader/CodeGen/Spirv/SpirvGenerator.cs:line 409
   at Ryujinx.Graphics.Shader.CodeGen.Spirv.SpirvGenerator.Generate(CodeGenContext context, StructuredProgramInfo info, Int32 funcIndex) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Shader/CodeGen/Spirv/SpirvGenerator.cs:line 156
   at Ryujinx.Graphics.Shader.CodeGen.Spirv.SpirvGenerator.Generate(StructuredProgramInfo info, CodeGenParameters parameters) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Shader/CodeGen/Spirv/SpirvGenerator.cs:line 127
   at Ryujinx.Graphics.Shader.Translation.TranslatorContext.Generate(IReadOnlyList`1 funcs, AttributeUsage attributeUsage, ShaderDefinitions definitions, ShaderDefinitions originalDefinitions, ResourceManager resourceManager, FeatureFlags usedFeatures, Byte clipDistancesWritten) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Shader/Translation/TranslatorContext.cs:line 372
   at Ryujinx.Graphics.Shader.Translation.TranslatorContext.Translate(FunctionCode[] functions, ResourceManager resourceManager, FeatureFlags usedFeatures, Byte clipDistancesWritten, Boolean asCompute) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Shader/Translation/TranslatorContext.cs:line 308
   at Ryujinx.Graphics.Shader.Translation.TranslatorContext.Translate(Boolean asCompute) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Shader/Translation/TranslatorContext.cs:line 211
   at Ryujinx.Graphics.Gpu.Shader.ShaderCache.TranslateShader(ShaderDumper dumper, GpuChannel channel, TranslatorContext context, Byte[] code, Boolean asCompute) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Gpu/Shader/ShaderCache.cs:line 777
   at Ryujinx.Graphics.Gpu.Shader.ShaderCache.GetGraphicsShader(ThreedClassState& state, ProgramPipelineState& pipeline, GpuChannel channel, GpuChannelPoolState& poolState, GpuChannelGraphicsState& graphicsState, ShaderAddresses addresses) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Gpu/Shader/ShaderCache.cs:line 305
   at Ryujinx.Graphics.Gpu.Engine.Threed.StateUpdater.UpdateShaderState() in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Gpu/Engine/Threed/StateUpdater.cs:line 1379
   at Ryujinx.Graphics.Gpu.Engine.Threed.DrawManager.DrawEnd(ThreedClass engine, Int32 firstIndex, Int32 indexCount, Int32 drawFirstVertex, Int32 drawVertexCount) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Gpu/Engine/Threed/DrawManager.cs:line 123
   at Ryujinx.Graphics.Gpu.Engine.GPFifo.GPFifoProcessor.Send(UInt64 gpuVa, Int32 offset, Int32 argument, Int32 subChannel, Boolean isLastCall) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Gpu/Engine/GPFifo/GPFifoProcessor.cs:line 255
   at Ryujinx.Graphics.Gpu.Engine.GPFifo.GPFifoProcessor.Process(UInt64 baseGpuVa, ReadOnlySpan`1 commandBuffer) in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Gpu/Engine/GPFifo/GPFifoProcessor.cs:line 82
   at Ryujinx.Graphics.Gpu.Engine.GPFifo.GPFifoDevice.DispatchCalls() in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Graphics.Gpu/Engine/GPFifo/GPFifoDevice.cs:line 206
   at Ryujinx.Ava.AppHost.<RenderLoop>b__101_1() in /home/runner/work/Ryujinx/Ryujinx/src/Ryujinx.Ava/AppHost.cs:line 922
   at System.Threading.Thread.StartCallback()
```

gdkchan commented:
> ClipDistance".
> 00:00:22.486 |W| GPU.MainThread Gpu Log: Shader translator: Invalid output "ClipDistance".
> 00:00:22.486 |W| GPU.MainThread Gpu Log: Shader translator: Invalid output "ClipDistance".
> 00:00:22.486 |W| GPU.MainThread Gpu Log: Shader translator: Invalid output "ClipDistance".

Should be fixed now, can you try again?

gssci commented:
Just FYI, or for anyone curious, this update for me fixes the rgb glowing glitch in Metroid prime. However there's still an error near doors and affected graphical elements, where black triangles start flashing on the screen.
Is this due to the current limitations you were mentioning? 

(https://github.com/Ryujinx/Ryujinx/files/12452076/Ryujinx_1.1.0%2Baae4557_2023-08-28_10-01-50.log)

https://github.com/Ryujinx/Ryujinx/assets/9202765/445e9d78-9c54-4a28-968b-d2729ed1abd2

![Screenshot 2023-08-28 at 10 10 25](https://github.com/Ryujinx/Ryujinx/assets/9202765/a7925f3a-febd-4c31-bb41-e50e175db67f)

gdkchan commented:
I think something was posted a while ago about weird triangles showing on Metroid Prime Remastered. The message seems to have disappeared for some reason, but I will answer it anyway.

It *might* be due to the features that are not implemented. In special this game uses a few things that are not working on macOS right now:
- Transform feedback from geometry shader.
- Primitives generated and transform feedback primitives written counters.
- Integer attribute type on indexed vertex shader inputs.

AFAIK those are used to render the shadow under the player (and other places). So maybe the black triangles are the shadow trying to render incorrectly due to those features not working properly. In any case, I made a change to disable trying to emulate transform feedback on geometry shaders entirely (since it was not working properly anyway), that seems to have stopped the weird black triangles (although I didn't play for too long, so maybe I was just "lucky").

gssci commented:
Amazing! Thank you! I deleted the comment myself, because I thought it might be off topic and a game specific issue. I will try the new build shortly 🎉
